### PR TITLE
HyperPlay Integration

### LINKED ISSUE #831: Hyperplay SDK Sample & Integration
**Summary**
Integrate the Hyperplay SDK into a sample scene and documentation

**Suggested features or ways to address the summary**
Follow the docs and provide a reference implementation of the Hyperplay SDK

**Motivation**
Hyperplay is a relevant component for access to an exciting distribution channel for games built with web3.untiy

**Acceptance Criteria**
- SIWE login via HyperPlay's desktop client launcher.
- Signing and executing transaction via HyperPlay's desktop client launcher.

**Additional Context**
[Ideation](https://www.notion.so/chainsafe/Hyperplay-CS-Gaming-ideation-d6015e28fae7415fbfca359ef3bbae65?pvs=4#a3564f4e2db74492b2584b6562fff198)
[Hyperplay Docs](https://docs.hyperplay.xyz/developer-docs/api)

### PR DESCRIPTION
**Pull request by rob1997**
closes #831 
SIWE implemented

### CODE REVIEW

**Review by sneakzttv — APPROVED** [review]

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                    Method = method,
+                    Params = parameters,
+                },
+                Chain = new Chain
+                {
+                    ChainId = chainConfig.ChainId,
+                },
+            });
+
+            string response = (await httpClient.PostRaw("localhost:9680/rpc", body, "application/json")).Response;
```
Comment (oleksandrchainsafe) [inline src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:66]: Who is responsible for stating up the server on the 9680 port? How can we be sure it's up and running at that point of time?

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+            try
+            {
+                var error = JsonConvert.DeserializeObject<HyperPlayError>(response);
+
+                throw new Web3Exception($"HyperPlay RPC request failed: {error.Message}");
+            }
+            catch (Exception)
```
Comment (oleksandrchainsafe) [inline src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:74]: I think we could catch a more specific json-something-exception here.

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayRequest.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public struct Chain
+    {
+        [JsonProperty("chainId")]
+        public string ChainId { get; set; }
+    }
+
+    /// <summary>
+    /// Error model for HyperPlay RPC responses.
+    /// </summary>
+    public struct HyperPlayError
```
Comment (oleksandrchainsafe) [inline src/ChainSafe.Gaming.HyperPlay/HyperPlayRequest.cs:35]: The file name doesn't really reflect the contents IMO. I would rather break this up into separate files under the "Dto" folder or rename this file to something like HyperPlayDtos.

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+            string hash = await hyperPlayProvider.Request<string>("eth_sendTransaction", transactionInput);
+
+            string hashPattern = @"^0x[a-fA-F0-9]{64}$";
+            if (!Regex.IsMatch(hash, hashPattern))
+            {
+                throw new Web3Exception($"incorrect txn response format {hash}");
+            }
+
+            return await hyperPlayProvider.Request<TransactionResponse>("eth_getTransactionByHash", hash);
```
Comment (oleksandrchainsafe) [inline src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs:64]: Don't know if this is applicable for the HyperPlayer provider, but we do polling with timeout for all the other TxExecutors (except MM as I just found out) for this call. The reason is, that it usually takes time for a transaction to be registered on a node, so when you try to get it by hash it fails at first.

Look at RpcProviderExtensions.GetTransaction() for more context.

File: scripts/data/published_dependencies.txt [hunk 5]
```diff
@@ -1,2 +1,2 @@
...
...CompilerServices.Unsafe;Microsoft.Extensions.DependencyInjection;System.Runtime.InteropServices.WindowsRuntime;Microsoft.Extensions.Logging.Abstractions;System.Security.Cryptography.Cng;Microsoft.IdentityModel.Abstractions;System.Text.Encodings.Web;Microsoft.IdentityModel.Logging;System.Text.Json;Microsoft.IdentityModel.Tokens;System.Threading.Channels;NBitcoin;System.Threading.Tasks.Extensions;Nethereum.ABI;WalletConnectSharp.Auth;Nethereum.Accounts;WalletConnectSharp.Common;WalletConnectSharp.Events;Nethereum.BlockchainProcessing;WalletConnectSharp.Core;Nethereum.Contracts;WalletConnectSharp.Crypto;Nethereum.Hex;Nethereum.JsonRpc.Client;WalletConnectSharp.Network.Websocket;Nethereum.JsonRpc.IpcClient;WalletConnectSharp.Network;Nethereum.JsonRpc.RpcClient;WalletConnectSharp.Sign;Nethereum.KeyStore;WalletConnectSharp.Storage;Nethereum.Merkle.Patricia;WalletConnectSharp.Web3Wallet;Nethereum.Merkle;Websocket.Client;Nethereum.Metamask;Nethereum.Siwe.Core;Nethereum.Siwe;Nethereum.UI;Nethereum.Unity.Metamask;Nethereum.Unity;ChainSafe.Gaming.MetaMask;ChainSafe.Gaming.MetaMask.Unity;ChainSafe.Gaming.InProcessTransactionExecutor.Unity;ChainSafe.Gaming.Marketplace;ChainSafe.Gaming.HyperPlay
```
[the remark this answers is not in the document]
Comment (rob1997) [reply scripts/data/published_dependencies.txt:1]: Emm I don't know, HyperPlay is just a Sign-in method for now, that means we would need extra packages for MM and WC too

[the remark this answers is not in the document]
Comment (rob1997) [reply scripts/data/published_dependencies.txt:1]: best to keep it in package for now imo

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        /// Connect to wallet via HyperPlay desktop client and return the account address.
+        /// </summary>
+        /// <returns>Signed-in account public address.</returns>
+        public async Task<string> Connect()
+        {
+            string[] accounts = await Request<string[]>("eth_accounts");
+
+            string account = accounts[0];
+
+            string hash = await Request<string>("personal_sign", "Sign-in with Ethereum.", account);
```
[the remark this answers is not in the document]
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:39]: the reason is to confirm sign-in but tbh I can avoid it

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs (hunk 1 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:66]: HyperPlay login requires HyperPlay desktop client to run and the desktop client runs this server by default

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs (hunk 4 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs:64]: yep, I'll test it a bit more to make sure

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayRequest.cs (hunk 3 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayRequest.cs:35]: gotcha, will change

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs (hunk 4 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs:64]: it seems to work consistently, maybe HyperPlay does their own polling until a node is confirmed

**Review by juans-chainsafe — APPROVED** [review]
Functionality working as expected in editor and Desktop build.

**Not working in WebGL.**
```
webgl5.framework.js.br:3 Web3Exception: HTTP.POST to localhost:9680/rpc responded with error: 
  at ChainSafe.Gaming.Web3.Core.Unity.UnityHttpClient.UnityWebRequestToNetworkResponse (UnityEngine.Networking.UnityWebRequest request)
```


I created this minor issue for Remember me functionality https://github.com/ChainSafe/web3.unity/issues/961

Screenshot 2024-05-16 at 10 03 03 AM

File: scripts/data/published_dependencies.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (oleksandrchainsafe) [reply scripts/data/published_dependencies.txt:1]: ok, let's keep it for now. I'm moving the WC integrations into their own packages, but this is very much an experiment at this point.

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs (hunk 1 above)
Comment (oleksandrchainsafe) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:64]: So we can be 100% sure we will reach it, right?

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs (hunk 4 above)
Comment (oleksandrchainsafe) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayTransactionExecutor.cs:64]: if it works, then no, keep it as it is

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs (hunk 1 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs:64]: yeah it's on their docs https://docs.hyperplay.xyz/developer-docs/api/sign-in-with-ethereum

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs (hunk 2 above)
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs]: yep, now replaced

File: src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rob1997) [reply src/ChainSafe.Gaming.HyperPlay/HyperPlayProvider.cs]: gotcha, now message is signed and hash is verified.

**Review by oleksandrchainsafe — APPROVED** [review]
✅

**Review by oleksandrchainsafe — APPROVED** [review]

**Pull request merged**